Base class for transported scalars

TobiKattmann opened the pull request:
## Proposed Changes
Hi all,

The general idea for this PR can be cited from @pcarruscag in #1226:
```
The ideia would be to create a CFVMSolverBase (in the style of the one for flow solvers) to cover:

    Turbulence solvers;
    Heat solver;
    Radiation solver;
    General passive/active scalar solvers;
    Eventual transition solvers;

The starting point would be the current base turbulence solver.
```
 
Right now the only things I did were (i.e. don't fear the large diff)
- create  CScalarSolver/Variable files/classes and scalar_convection/diffusion/sources files and fill them with the contents of the CTurb* or turb_* files.
- Make CScalarSolver/Variable the parent class of CTurbSolver/Variable
- Delete pretty much everything from CTurb* or turb_* files except for some constructors

So far it compiled and some local regression testing showed no problems, lets see what the CI pipeline has to say about this

Now with this additional class introduced the actual works begins:
- figure out what is a reasonable interface in CScalar* to unify transported species and turb-models (and the other solvers probably)
- code this common interface and push all turb-specific things back into the respective files (that is why I didn't `git mv` )

In order to keep this somewhat limited I would focus on porting the Turbsolvers to the new structure (however that might look, but looking at #1223 is what I'll do first) and maybe a simple transported passive scalar or so. We want the turb solver to be integrated under that Scalar-class which is arguably what most/all people use, that's why I want to have in this work at the beginning. We also have quite a few RANS Testcases so it is directly visible if something in the actual computation is changed.

In case there are comments, ideas foreseeable stepping stones from experience please let me know here.

## Related Work
#777 by @economon which is the initial push towards this separate scalar solver class for additionally transported scalars (not merged)
#1223 by @danielmayer @bigfooted  which builds up on #777 and extends it to its needs (but does not support the turb solvers e.g.)
#1226 by @oleburghardt - In this PR is a discussion on creating the structure which this (and following) PR('s) try to achieve
#1332 by @mheimgartner - implementing multi component flow (or analytic mixing properties of said multicomp flow). That PR is built upon #1223 but will merge this PR (once it is done) and throw out all additional changes in order to be merged into develop without relying on feature_flamelet



## PR Checklist

- [x] I am submitting my contribution to the develop branch.
- [x] My contribution generates no new compiler warnings (try with the '-Wall -Wextra -Wno-unused-parameter -Wno-empty-body' compiler flags, or simply --warnlevel=2 when using meson).
- [x] My contribution is commented and consistent with SU2 style.
- [ ] I have added a test case that demonstrates my contribution, if necessary.
- [ ] I have updated appropriate documentation (Tutorials, Docs Page, config_template.cpp) , if necessary.

pcarruscag commented on SU2_CFD/src/numerics/scalar/scalar_sources.cpp:
```diff
@@ -0,0 +1,930 @@
+/*!
+ * \file turb_sources.cpp
+ * \brief Implementation of numerics classes for integration of
+ *        turbulence source-terms.
```
Again :+1: for taking the plunge on this.
I would not have moved the sources, I think these will always be very specific to each derived solver.
It would be better for navigation to keep it in the turbulence folder.
TBH I would not have moved anything that was not immediately necessary (all numerics in this case) because if someone comes along with changes to those classes....

TobiKattmann replied:
My motivation to move everything was more of a test/first-step to see if I break anything with that -> inserting the CScalar layer and moving things there. Turbulent sources will certainly move back and everything else specific will move back at some point.

> code this common interface and push all turb-specific things back into the respective files (that is why I didn't git mv )

I still have to figure out how a suitable generalized interface looks like and based on that I will reject the temporary residence permits for some stuff (e.g. all turb sources) 👍

TobiKattmann replied:
But by all means, I highly appreciate all sorts of comments as I there is lots of things to do wrong so I am glad for additional brains to look at this 🧠 👀

pcarruscag replied:
Sure but you only need to worry about the solver and variables. The numerics are fairly independent (for now).
Minimize the number of changes per PR is all I'm saying. I think that's a guideline we all endorse :)

TobiKattmann replied:
> TBH I would not have moved anything that was not immediately necessary (all numerics in this case) because if someone comes along with changes to those classes....

well, yes that happenend ... should have listened directly

TobiKattmann commented:
Still WIP but with that in the title the regression status behind each commit is always "unfinished" because WIP-check-bot does that. Now that we have Draft-PR I guess we dont WIP anymore tbh

pcarruscag commented:
It helps me with the manual edit of the release notes, to know what was not merged yet.

TobiKattmann commented:
Things done:
- Create CScalarSolver/Variable classes that now sits between CSolver/Variable and CTurbSolver/Variable
- Move variables/methods to CScalar-land for which one can reasonably assume that they will be useful for future children of the CScalar-family
- That implies, that the rest remains in Turbulence land mostly(!) untouched
- Mostly, because I renamed (Set)TurbVar[-,Gradient] variables/methods in CNumerics which result in changing that all over turb_diff/conv/sources. They hear by the name (Set)ScalarVar* now
And that's it for now. I added some personal comments in the code here and there that will be removed.

Next up (coming to a PR near you soon):
- add CSpeciesSolver/Variable with appropriate Upwind/Viscous_Residual and Boundary conditions
- The first step will be a truly Passive Species transport but then ...
- ...introduce the mixture properties of @mheimgartner from #1332 i.e. make the transport equation 'active'. This comes along with some more necessary changes to CConfig. I am not sure yet what is the best order of tackling that. We'll see...
- @mheimgartner work concentrates on 1 additional transport equation. The idea is of course to make it as general as possible but it might be wiser to just go with 1 for the beginning.
- For that setup (2 components mixing) we'll want to get validated gradients going
- The more general way will be needed to merge with @danielmayer #1223 but that would be the step after the above mentioned

@bigfooted that i what I currently have in mind but feel free to chime in here.

pcarruscag commented on SU2_CFD/include/variables/CTurbVariable.hpp:
```diff
@@ -27,22 +27,19 @@
-class CTurbVariable : public CVariable {
+class CTurbVariable : public CScalarVariable {
 protected:
   VectorType muT;         /*!< \brief Eddy viscosity. */
```
The solvers see the nodes as CScalarVariable, but the methods to access muT are in CTurbVariable.
The compiler does not know that the type of nodes will be CTurbVariable at runtime and so it cannot de-virtualize the access to variables that are not in CScalarVariable.

By making CScalarSolver a class template you can instantiate it with different types for "nodes", i.e. in practice you end up with multiple CScalarSolver "tuned" for each of the derived classes. CScalarSolver<CTurbVariable>, CScalarSolver<CHeatVariable> (which are not the same solver anymore, but that is fine because we want this as a building block.

TobiKattmann replied:
I did a little debugging on restarts over at feature_flamelet and feature_multicomp, so this will be a bit delayed 🐌 sorry for that

TobiKattmann replied:
Should I go for a similar approach as `CFVMFlowSolverBase.hpp/inl` with the explicit instantiation of the Solver i need in the derived solvers? i.e. `template class CScalarSolver<CTurbVariable>;` at the top of CTurbVariable.cpp?

alternatively I could include CScalarSolver.cpp (or rename to .inl) at the end of its hpp?

pcarruscag replied:
Do it like CFVMFlowSolverBase.hpp and .inl please.

TobiKattmann replied:
Implemented in ca16276d50b2dc99ea40c511aa6df70b141b986c Let me know if that is what you think should be done

TobiKattmann replied:
Just for my own documentation: [template use in implementation file](https://stackoverflow.com/questions/18186878/invalid-use-of-template-name-without-an-argument-list) / [templates implemented in header files](https://stackoverflow.com/questions/495021/why-can-templates-only-be-implemented-in-the-header-file)

pcarruscag commented on SU2_CFD/src/solvers/CScalarSolver.cpp:
```diff
@@ -0,0 +1,958 @@
+  delete nodes;
+
+}
+
```
Indeed that is annoying, I think I've put su2double in front of all flow solver vars because of that. Otherwise the problem comes back, or we have to disable clang-format around certain sections.

pcarruscag replied:
> The 1 space indentation of protected seems weird as well

It's the price to pay for peace of mind with formatting, it's easier to get used to that than to every personal style in SU2 :laughing:

pcarruscag commented on SU2_CFD/src/solvers/CScalarSolver.cpp:
```diff
@@ -0,0 +1,958 @@
+  /*--- Roughness heights. ---*/
+  if (config->GetKind_Turb_Model() == SA)
+    numerics->SetRoughness(geometry->nodes->GetRoughnessHeight(iPoint),
+                           geometry->nodes->GetRoughnessHeight(jPoint));
```
Something along the lines of what we are doing to compute time-steps in the flow solver, to keep it nice and speedy.

TobiKattmann replied:
Implemented in https://github.com/su2code/SU2/pull/1330/commits/3e975d3a69f495dfc7507246c816012b1b520fde let me know whether this is what you had in mind 👍

pcarruscag commented on SU2_CFD/src/solvers/CScalarSolver.cpp:
```diff
@@ -0,0 +1,958 @@
+            density_old = flowNodes->GetSolution_Old(iPoint,0);
+
+          for (unsigned short iVar = 0; iVar < nVar; iVar++) {
+            nodes->AddConservativeSolution(iPoint, iVar,
```
Ah... yet another variant, well we can always override those very special cases. It is probably better to have one or two exceptions than to shoe-horn every solver into an interface that will be inefficient in 90% of cases (every extra variable is a little more memory for AD for example).

TobiKattmann replied:
I added a `bool Conservative` to Cscalarsolver as class data and used the more general `AddConservativeSolution` for both, the clipping limits are default set to su2doule::min/max in case a solver does not use this (e.g. SA)

pcarruscag commented on SU2_CFD/include/solvers/CScalarSolver.hpp:
```diff
@@ -48,11 +48,13 @@ class CScalarSolver : public CSolver {
 
-  su2double lowerlimit[MAXNVAR] = {0.0}, /*!< \brief contains lower limits for turbulence variables. */
-      upperlimit[MAXNVAR] = {0.0};       /*!< \brief contains upper limits for turbulence variables. */
+  su2double lowerlimit[MAXNVAR] = {std::numeric_limits<su2double>::lowest()}; /*!< \brief contains lower limits for turbulence variables. Note that ::min() returns the smallest positive value for floats. */
+  su2double upperlimit[MAXNVAR] = {std::numeric_limits<su2double>::max()}; /*!< \brief contains upper limits for turbulence variables. */
```
This does not do what you think it does :) after the first you have zeros.

TobiKattmann replied:
hm, yes. it didn't break on me because SA only has 1 var (but some solver in the future might have more and does not explicitly set those). I'll fix it

TobiKattmann replied:
Now it is done via a loop in the CScalarVar ctor, so each derived solver can overwrite that

TobiKattmann commented on SU2_CFD/include/solvers/CScalarSolver.hpp:
```diff
@@ -48,7 +48,7 @@ class CScalarSolver : public CSolver {
   unsigned long omp_chunk_size; /*!< \brief Chunk size used in light point loops. */
 
-  su2double lowerlimit[MAXNVAR] = {std::numeric_limits<su2double>::min()}; /*!< \brief contains lower limits for turbulence variables. */
+  su2double lowerlimit[MAXNVAR] = {std::numeric_limits<su2double>::lowest()}; /*!< \brief contains lower limits for turbulence variables. Note that ::min() returns the smallest positive value for floats. */
```
I wasn't aware of that. Maybe that is a noteworthy info.

TobiKattmann commented on SU2_CFD/src/solvers/CScalarSolver.cpp:
```diff
@@ -450,12 +448,17 @@ void CScalarSolver::SetResidual_DualTime(CGeometry* geometry, CSolver** solver_c
   su2double Residual_GCL;
 
+  const int nvar = nVar;
+  auto FactorTimesVector = [nvar](su2double factor, su2double* vec) {
+    for (int iVar = 0; iVar < nvar; iVar++) vec[iVar] = factor * vec[iVar];
+  };
```
Maybe this is already implemented somewhere centrally? (Note that I wanted to avoid the term Scalar ...)

TobiKattmann replied:
nvm, removed that

pcarruscag commented on SU2_CFD/src/solvers/CScalarSolver.cpp:
```diff
@@ -372,40 +373,26 @@ void CScalarSolver::CompleteImplicitIteration(CGeometry* geometry, CSolver** sol
+    SU2_OMP_FOR_STAT(omp_chunk_size)
+    for (unsigned long iPoint = 0; iPoint < nPointDomain; iPoint++) {
+      /*--- Multiply the Solution var with density to get the conservative transported quantity, if necessary. ---*/
+      if (Conservative) {
+        density = flowNodes->GetDensity(iPoint);
+        density_old = compressible ? flowNodes->GetSolution_Old(iPoint, 0) : density;
+      }
```
That will work I guess, but it throws away optimizations.
Better use if (conservative) (outside of the loop over variables) to call either AddConservativeSolution or AddClippedSolution.

TobiKattmann replied:
Ok I removed that but I just kept AddConservativeSolution renamed to AddClippedSolution and made the factor between Solution and Transported variable default to 1. Thus the difference between the if(Conservative) ...} else {..} is really just the density. See 817186f826785e43d5aa8ae397846b39698dabbd . Let me know what you think

pcarruscag commented on SU2_CFD/src/solvers/CScalarSolver.cpp:
```diff
@@ -0,0 +1,811 @@
+                (Density_nP1 * U_time_nP1[iVar] - Density_n * U_time_n[iVar]) * (Volume_nP1 / TimeStep);
+          if (second_order)
+            LinSysRes(iPoint, iVar) +=
+                (Density_nP1 * U_time_nP1[iVar] - Density_n * U_time_n[iVar]) * (3.0 * Volume_nP1 / (2.0 * TimeStep)) +
+                (Density_nM1 * U_time_nM1[iVar] - Density_n * U_time_n[iVar]) * (Volume_nM1 / (2.0 * TimeStep));
+        }
+
+      } else {
+        for (iVar = 0; iVar < nVar; iVar++) {
+          if (first_order) LinSysRes(iPoint, iVar) += (U_time_nP1[iVar] - U_time_n[iVar]) * (Volume_nP1 / TimeStep);
```
and here

pcarruscag commented on SU2_CFD/src/solvers/CScalarSolver.cpp:
```diff
@@ -0,0 +1,811 @@
+
+    /*--- Skip flow variables ---*/
+
+    unsigned short skipVars = 0;
+
+    if (nDim == 2) skipVars += 6;
+    if (nDim == 3) skipVars += 8;
```
```suggestion
    unsigned short skipVars = nDim + solver[MESH_0][FLOW_SOL]->GetnVar();
```

pcarruscag commented on SU2_CFD/src/solvers/CTurbSASolver.cpp:
```diff
@@ -1591,3 +1588,54 @@ void CTurbSASolver::SetUniformInlet(const CConfig* config, unsigned short iMarke
+   SA_NEG model is more robust due to allowing for negative nu_tilde,
+   so the under-relaxation is not applied to that variant. */
+
+  bool sa_model = ((config->GetKind_Turb_Model() == SA)        ||
+                   (config->GetKind_Turb_Model() == SA_E)      ||
+                   (config->GetKind_Turb_Model() == SA_COMP)   ||
+                   (config->GetKind_Turb_Model() == SA_E_COMP));
```
```suggestion
  if (config->GetKind_Turb_Model() == SA_NEG) return;
```

TobiKattmann replied:
Although I am not fully sold on reversing these kinds of logics, who knows what is about to come

pcarruscag commented on SU2_CFD/src/solvers/CTurbSASolver.cpp:
```diff
@@ -1591,3 +1588,54 @@ void CTurbSASolver::SetUniformInlet(const CConfig* config, unsigned short iMarke
+  for (unsigned long iPoint = 0; iPoint < nPointDomain; iPoint++) {
+
+    localUnderRelaxation = 1.0;
+    if (sa_model) {
+      for (unsigned short iVar = 0; iVar < nVar; iVar++) {
+
+        /* We impose a limit on the maximum percentage that the
+          turbulence variables can change over a nonlinear iteration. */
+
+        const unsigned long index = iPoint * nVar + iVar;
+        su2double ratio = fabs(LinSysSol[index]) / (fabs(nodes->GetSolution(iPoint, iVar)) + EPS);
+        if (ratio > allowableRatio) {
+          localUnderRelaxation = min(allowableRatio / ratio, localUnderRelaxation);
+        }
+
+      }
+    }
```
```suggestion
    su2double ratio = fabs(LinSysSol[iPoint]) / (fabs(nodes->GetSolution(iPoint, 0)) + EPS);
    /* We impose a limit on the maximum percentage that the
      turbulence variables can change over a nonlinear iteration. */
    if (ratio > allowableRatio) {
      localUnderRelaxation = min(allowableRatio / ratio, localUnderRelaxation);
    }
```

TobiKattmann replied:
yep, sa has indeed only 1 variable 👍

TobiKattmann commented on SU2_CFD/src/solvers/CScalarSolver.cpp:
```diff
@@ -0,0 +1,811 @@
+            else
+              Density_n = flowNodes->GetSolution_time_n(iPoint, 0);
+
+            for (iVar = 0; iVar < nVar; iVar++) LinSysRes(iPoint, iVar) += Density_n * U_time_n[iVar] * Residual_GCL;
+          } else {
+            for (iVar = 0; iVar < nVar; iVar++) LinSysRes(iPoint, iVar) += U_time_n[iVar] * Residual_GCL;
```
👍 yes, I was a bit lazy for the dynamic grid portion

TobiKattmann commented on SU2_CFD/include/variables/CScalarVariable.hpp:
```diff
@@ -0,0 +1,77 @@
+ */
+class CScalarVariable : public CVariable {
+ protected:
+  MatrixType HB_Source; /*!< \brief Harmonic Balance source term. */
```
👍 yep they were in the SA-Variable, so they weren't accessible for the SST solver as well

TobiKattmann commented on SU2_CFD/src/solvers/CScalarSolver.cpp:
```diff
@@ -0,0 +1,799 @@
+      Read_SU2_Restart_ASCII(geometry[MESH_0], config, restart_filename);
+    }
+
+    /*--- Skip flow variables ---*/
+
+    unsigned short skipVars = nDim + solver[MESH_0][FLOW_SOL]->GetnVar();
```
`CScalarSolver::LoadRestart` will work like this only in a 'choose one' fashion. For Turb and Species Transport There needs to be a clear hierarchy (which is given by the ouptut of course). Although not strictly necessary (I think), keeping the order of how the linear systems are solved might be good for OCD-patients. 
I personally would keep this method right here because the majority stays the same. idk have a `skipVars = DetermineSkpVars(config)` method

TobiKattmann commented on SU2_CFD/src/solvers/CScalarSolver.cpp:
```diff
@@ -0,0 +1,799 @@
+
+  /*--- MPI solution and compute the eddy viscosity ---*/
+
+  solver[MESH_0][TURB_SOL]->InitiateComms(geometry[MESH_0], config, SOLUTION);
+  solver[MESH_0][TURB_SOL]->CompleteComms(geometry[MESH_0], config, SOLUTION);
+
+  solver[MESH_0][FLOW_SOL]->Preprocessing(geometry[MESH_0], solver[MESH_0], config, MESH_0, NO_RK_ITER,
+                                          RUNTIME_FLOW_SYS, false);
+  solver[MESH_0][TURB_SOL]->Postprocessing(geometry[MESH_0], solver[MESH_0], config, MESH_0);
```
The other things that is solver specific part is especially what to Pre/Postprocess after solution loading to stay consistent with the 'normal' iteration. For species transport  with turb this gets a tiny bit more involved and with 4 solvers and more cross-dependencies one might have to do quite a bunch of calls to get everything updated.
(and the SOLVER_IDX needs to be specialized)

TobiKattmann commented on SU2_CFD/src/solvers/CTurbSASolver.cpp:
```diff
@@ -289,6 +286,24 @@ void CTurbSASolver::Postprocessing(CGeometry *geometry, CSolver **solver_contain
+  /*--- Define an object to set solver specific numerics contribution. ---*/
+  struct SolverSpecificNumerics {
+
+    FORCEINLINE void operator() (const CGeometry& geometry, CNumerics& numerics, CConfig& config, const CScalarVariable& nodes, unsigned long iPoint, unsigned long jPoint) const {
+      /*--- Roughness heights. ---*/
+      if (config.GetKind_Turb_Model() == SA)
+        numerics.SetRoughness(geometry.nodes->GetRoughnessHeight(iPoint), geometry.nodes->GetRoughnessHeight(jPoint));
+    }
```
I know this is probably redundant, but I kept it for know to see whether the regression test pass. (this commit shouldn't change anything logically)
Is there any reason why surface roughness and other SA-variants should not work together

pcarruscag replied:
It is not implemented for other types of SA source, unfortunately.
Is there an incompatibility @koodlyakshay ?
The work of @suargi might fix this.

koodlyakshay replied:
I know the roughness model works with the noft2 version (which we have in SU2 now) and the standard version of SA. I have not seen it used with other versions anywhere.

TobiKattmann replied:
Ok thanks, so the condition `if SA` can stay 👍

pcarruscag commented on SU2_CFD/include/solvers/CScalarSolver.hpp:
```diff
@@ -0,0 +1,294 @@
+    numerics->SetScalarVar(nodes->GetSolution(iPoint), nodes->GetSolution(jPoint));
+    numerics->SetScalarVarGradient(nodes->GetGradient(iPoint), nodes->GetGradient(jPoint));
+
+    /*--- Call Numerics contribution which are Solver-Specifc. Implemented in the caller: Viscous_Residual.  ---*/
+
+    SolverSpecificNumerics(*geometry, *numerics, *config, *nodes, iPoint, jPoint);
```
I would have passed just iPoint / jPoint because the caller has everything already.

pcarruscag commented on SU2_CFD/include/solvers/CScalarSolver.hpp:
```diff
@@ -0,0 +1,294 @@
+   * \param[in] config - Definition of the particular problem.
+   * \note Calls a generic implementation after defining a SolverSpecificNumerics object.
+   */
+  inline virtual void Viscous_Residual(unsigned long iEdge, CGeometry* geometry, CSolver** solver_container,
+                                       CNumerics* numerics, CConfig* config) {}
```
You can give it default behavior by passing an empty solver-specific object.
If you see strong reasons against that, make this pure virtual so that no one forgets this needs to be implemented in derived classes.

TobiKattmann replied:
I went with the empty solver specific implementation instead of pure virtual. I guess as of right now all solvers will need to their own stuff so one could pure virtual imo but maybe there will be a gneric Diffusivity at one point and then this might come in handy
Done in https://github.com/su2code/SU2/pull/1330/commits/1d8f520a9da2907b105b3f47f2fcd05e75d424c9

pcarruscag commented on SU2_CFD/src/solvers/CTurbSASolver.cpp:
```diff
@@ -289,6 +286,24 @@ void CTurbSASolver::Postprocessing(CGeometry *geometry, CSolver **solver_contain
+  /*--- Define an object to set solver specific numerics contribution. ---*/
+  struct SolverSpecificNumerics {
+
+    FORCEINLINE void operator() (const CGeometry& geometry, CNumerics& numerics, CConfig& config, const CScalarVariable& nodes, unsigned long iPoint, unsigned long jPoint) const {
+      /*--- Roughness heights. ---*/
+      if (config.GetKind_Turb_Model() == SA)
+        numerics.SetRoughness(geometry.nodes->GetRoughnessHeight(iPoint), geometry.nodes->GetRoughnessHeight(jPoint));
+    }
+
+  } SolverSpecificNumerics;
```
e.g. if you reduce the signature to iPoint/jPoint, also note that you would need to template the "nodes" of the struct for when you template the scalar solver. If you let the caller "capture" them that is not necessary.
```suggestion
  auto SolverSpecificNumerics = [&](unsigned long iPoint, unsigned long jPoint) {
    if (config->GetKind_Turb_Model() == SA)
      numerics->SetRoughness(geometry->nodes->GetRoughnessHeight(iPoint),
                             geometry->nodes->GetRoughnessHeight(jPoint));
  };
```

TobiKattmann replied:
👍 ok so a lambda function instead of struct. That is indeed nicer to just capture everything that is used inside by reference

pcarruscag replied:
yep, if all you need is "operator ()", lambdas are simpler I think.

pcarruscag commented on SU2_CFD/src/solvers/CTurbSSTSolver.cpp:
```diff
@@ -280,6 +275,25 @@ void CTurbSSTSolver::Postprocessing(CGeometry *geometry, CSolver **solver_contai
+
+    FORCEINLINE void operator() (const CGeometry& geometry, CNumerics& numerics, CConfig& config, const CScalarVariable& nodes, unsigned long iPoint, unsigned long jPoint) const {
+      /*--- Menter's first blending function (only SST)---*/
+      if ((config.GetKind_Turb_Model() == SST) || (config.GetKind_Turb_Model() == SST_SUST))
+        numerics.SetF1blending(nodes.GetF1blending(iPoint), nodes.GetF1blending(jPoint));
```
```suggestion
      numerics.SetF1blending(nodes.GetF1blending(iPoint), nodes.GetF1blending(jPoint));
```

TobiKattmann replied:
also done in https://github.com/su2code/SU2/pull/1330/commits/3e86076ad4641710db97f5aa38cd23af11b419ae

pcarruscag reviewed (commented):
Yep that's the idea, nice :+1: 
Just a few things that may simplify the process a bit.

pcarruscag commented on SU2_CFD/include/variables/CScalarVariable.hpp:
```diff
@@ -0,0 +1,93 @@
+   * \param[in] iVar - Index of the variable.
+   * \param[in] source - Value of the harmonic balance source term. for the index <i>iVar</i>.
+   */
+  inline void SetHarmonicBalance_Source(unsigned long iPoint, unsigned long iVar, su2double source) override { HB_Source(iPoint,iVar) = source; }
```
```suggestion
  inline void SetHarmonicBalance_Source(unsigned long iPoint, unsigned long iVar, su2double source) final { HB_Source(iPoint,iVar) = source; }
```

pcarruscag commented on SU2_CFD/include/variables/CScalarVariable.hpp:
```diff
@@ -0,0 +1,93 @@
+   * \param[in] iVar - Index of the variable.
+   * \return Value of the harmonic balance source term for the index <i>val_var</i>.
+   */
+  inline su2double GetHarmonicBalance_Source(unsigned long iPoint, unsigned long iVar) const override { return HB_Source(iPoint,iVar); }
```
```suggestion
  inline su2double GetHarmonicBalance_Source(unsigned long iPoint, unsigned long iVar) const final { return HB_Source(iPoint,iVar); }
```

pcarruscag commented on SU2_CFD/include/variables/CScalarVariable.hpp:
```diff
@@ -0,0 +1,93 @@
+
+/*!
+ * \class CScalarVariable
+ * \brief Base class for defining the variables of the turbulence model.
+ * \ingroup Turbulence_Model
+ * \author A. Bueno.
```
```suggestion
 * \brief Base class for defining the shared variables of scalar solvers.
```

pcarruscag commented on SU2_CFD/include/variables/CScalarVariable.hpp:
```diff
@@ -0,0 +1,93 @@
+/*!
+ * \file CScalarVariable.hpp
+ * \brief Base class for defining the variables of the turbulence model.
```
```suggestion
 * \brief Base class for defining the shared variables of scalar solvers.
```

pcarruscag commented on SU2_CFD/src/solvers/CTurbSolver.cpp:
```diff
@@ -30,9 +30,9 @@
 #include "../../../Common/include/toolboxes/geometry_toolbox.hpp"
 
```
```suggestion
#include "../../include/solvers/CScalarSolver.inl"

```

pcarruscag commented on SU2_CFD/src/solvers/CTurbSSTSolver.cpp:
```diff
@@ -25,11 +25,15 @@
 #include "../../../Common/include/parallelization/omp_structure.hpp"
 #include "../../../Common/include/toolboxes/geometry_toolbox.hpp"
 
+/*--- Explicit instantiation of the parent class of CTurbSolver,
+ *    to spread the compilation over multiple cpp files. ---*/
+template class CScalarSolver<CTurbVariable>;
```
```suggestion
```

pcarruscag commented on SU2_CFD/src/solvers/CTurbSSTSolver.cpp:
```diff
@@ -25,11 +25,15 @@
  * License along with SU2. If not, see <http://www.gnu.org/licenses/>.
  */
 
+#include "../../include/solvers/CScalarSolver.inl"
```
```suggestion
```

TobiKattmann commented on SU2_CFD/src/solvers/CTurbSASolver.cpp:
```diff
@@ -1591,3 +1606,43 @@ void CTurbSASolver::SetUniformInlet(const CConfig* config, unsigned short iMarke
+  }
+  END_SU2_OMP_FOR
+
+}
```
```suggestion
}

```

TobiKattmann commented on SU2_CFD/src/solvers/CTransLMSolver.cpp:
```diff
@@ -25,18 +25,22 @@
        DOI: https://doi.org/10.2514/1.42362 ---*/
        
-CTransLMSolver::CTransLMSolver(void) : CTurbSolver() {}
+CTransLMSolver::CTransLMSolver(void) : CTurbSolver(true) {} ///Note: TransLM seems to rho*gamma, rho*Re_sigma as Solution variables
```
```suggestion
// Note: TransLM seems to use rho*gamma, rho*Re_sigma as Solution variables, thus Conservative=true
CTransLMSolver::CTransLMSolver(void) : CTurbSolver(true) {}
```

TobiKattmann commented on SU2_CFD/src/solvers/CTransLMSolver.cpp:
```diff
@@ -25,11 +25,15 @@
  */
 
-
+#include "../../include/solvers/CScalarSolver.inl"
```
```suggestion
```

TobiKattmann commented on SU2_CFD/include/solvers/CScalarSolver.hpp:
```diff
@@ -0,0 +1,301 @@
+ * \brief Main class for defining a scalar solver.
+ * \tparam VariableType - Class of variable used by the solver inheriting from this template.
+ */
+template <class TVariable>
```
```suggestion
template <class VariableType>
```

TobiKattmann commented on SU2_CFD/include/solvers/CScalarSolver.hpp:
```diff
@@ -35,10 +34,10 @@
+ * \brief Main class for defining a scalar solver.
+ * \tparam TVariable - Class of C*Variable of the specific solver
  */
+template<class TVariable>
```
I forget to add a this to the batch ... and now had to re-add this suggestions because it is not visible in the changed files any more 🤷

TobiKattmann commented on SU2_CFD/include/solvers/CScalarSolver.inl:
```diff
@@ -30,9 +29,11 @@
 
-CScalarSolver::CScalarSolver(bool conservative) : CSolver(), Conservative(conservative) {}
+template<class TVariable>
+CScalarSolver<TVariable>::CScalarSolver(bool conservative) : CSolver(), Conservative(conservative) {}
```
I went with <VariableType> instead of <V> to use the same as in the hpp. Maybe it is a bit more consistent this way idk

TobiKattmann replied:
See  6f8e69a

pcarruscag reviewed (approved):
Awesome work :+1: can't wait to see the crash diet that some solvers are about to go on.

TobiKattmann commented:
Thanks @pcarruscag for making this PR a lot better.

@bigfooted @danielmayer @mheimgartner 
To layout future steps as discussed after this merge: 
1. Merge these changes into feature_flamelet (and then feature_multicomp). Therefore the `CScalar*` things that are currently  present in these branches will be renamed to CLegacyScalar* to not overwrite files. This has the advantage that those branches are kept up-to-date with develop.
2. Implement a rudimentary (passive) CSpeciesSolver (specifically does not contain multicomponent mixing from #1332 ) and merge that. Finding a good (validation) Testcase would be great here.
3. Extend the CSpeciesSolver by #1332 mixing models. Either by cutting away all the flamelet specific parts or by porting the additions to a new PR.

Additionally the heat and radiation solver should be ported into this new structure, but that is independent work.
